File: source/python/125. Valid Palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = s.lower()
        characters = [i for i in s if i.isalnum()]
        return characters == characters[::-1]


def test():
    s = Solution()

    assert s.isPalindrome("A man, a plan, a canal: Panama") == True
    assert s.isPalindrome("race a car") == False
    assert s.isPalindrome(" ") == True

    print("All tests passed!")


if __name__ == "__main__":
    test()

File: source/cpp/125. Valid Palindrome.cpp
#include <string>
#include <string_view>


class Solution {
public:
    bool isPalindrome(std::string_view s) {
        if (s.empty()) return true;

        auto left = s.begin();
        auto right = s.end() - 1;
        while (left < right) {
            if (!std::isalnum(*left)) {
                ++left;
                continue;
            }
            if (!std::isalnum(*right)) {
                --right;
                continue;
            }
            if (std::tolower(*left) != std::tolower(*right)) {
                return false;
            }
            ++left;
            --right;
        }
        
        return true;
    }
};